Bump unconfirmed ancestor transactions to target feerate

murchandamus opened the pull request:
Includes some commits to address follow-ups from #27021: https://github.com/bitcoin/bitcoin/pull/27021#issuecomment-1554675156

Reduces the effective value of unconfirmed UTXOs by the fees necessary to bump their ancestor transactions to the same feerate.

While the individual UTXOs always account for their full ancestry before coin-selection, we can correct potential overestimates with a second pass where we establish the ancestry and bump fee for the whole input set collectively.

Fixes #9645 
Fixes #9864 
Fixes #15553

glozow committed:
[node] interface to get bump fees

fanquake commented on src/node/mini_miner.cpp:
```diff
@@ -0,0 +1,219 @@
+    for (const auto& txiter : cluster) {
+        const auto& txid = txiter->GetTx().GetHash();
+        // Cache descendants for future use. Unlike the real mempool, a descendant MockMempoolEntry
+        // will not exist wihout its ancestor MockMempoolEntry, so these sets won't be invalidated.
```
```suggestion
        // will not exist without its ancestor MockMempoolEntry, so these sets won't be invalidated.
```

fanquake commented on test/functional/wallet_spend_unconfirmed.py:
```diff
@@ -0,0 +1,297 @@
+#!/usr/bin/env python3
```
```bash
File "test/functional/wallet_spend_unconfirmed.py" contains a shebang line, but has the file permission 644 instead of the expected executable permission 755. Do "chmod 755 test/functional/wallet_spend_unconfirmed.py" (or remove the shebang line).
ERROR: There were 1 failed tests in the lint-files.py lint test. Please resolve the above errors.
```

This also needs to be added to the list of tests in `test_runner.py`. Which should deal with:
```bash
�[1mWARNING!�[0m The following scripts are not being run: ['wallet_spend_unconfirmed.py']. Check the test lists in test_runner.py.
```

fanquake reviewed (commented):
Concept ACK

Xekyo commented:
Thanks @fanquake, I fixed the two issues.

I also added a test for a transaction using `subtractfeefromamount`

glozow reviewed (commented):
HUGE Concept ACK obviously 🥳 🥳 

~This needs to obey -maxtxfee, so I'd suggest that coin selection keeps track of what ancestors the tx has and how much of the fees is ancestors vs this transaction. And we'd want a test where the target feerate is just below maxtxfee and there are low-feerate ancestors to bump, and so `total fees paid / size of this tx` is higher than maxtxfee, but actually you're just paying to bump.~

~Also, given that there is a (small) chance of overpayment, it would be good to check that -maxtxfee protects against something drastic. For example, a test where all the coins share 1 large parent? It will overestimate and -maxtxfee should make sure the tx isn't sent?~ edit: this is totally wrong

glozow commented on src/txmempool.cpp:
```diff
@@ -1197,3 +1216,58 @@ void CTxMemPool::SetLoadTried(bool load_tried)
+                        // transactions and skip.
+                        std::swap(cluster[i+1], cluster.back());
+                        cluster[i+1] = parent_it;
+                        ++i; 
```
Sorry, lint error is my fault here

```suggestion
                        ++i;
```

jonatack commented on src/txmempool.cpp:
```diff
@@ -1197,3 +1216,58 @@ void CTxMemPool::SetLoadTried(bool load_tried)
+                        ++i; 
+                    } else {
+                        cluster.push_back(parent_it);
+                        // we still need to process this 
```
whitespace linter
```suggestion
                        // we still need to process this
```

jonatack commented on src/node/mini_miner.h:
```diff
@@ -0,0 +1,104 @@
+    int64_t vsize_with_ancestors;
+    const CTransaction& tx;
+
+public:
```
All struct members are public by default, so can either drop `public` here or make the struct a class if the data members above are intended to be private.

jonatack reviewed (commented):
Concept ACK

jonatack commented on src/node/mini_miner.h:
```diff
@@ -0,0 +1,104 @@
+    int64_t vsize_with_ancestors;
+    const CTransaction& tx;
+
+public:
```
All struct members are public by default, so can either drop `public` here (and removed the getters), or make the struct a class if any of the data members above are intended to be private (for instance, `tx` and `fee_individual` have public getters and could be private, or just be public and drop the getter).

At first look it seems some of the data members need to be public, in which case their getters should be used or removed. This builds:

```diff
-struct MockMempoolEntry {
+class MockMempoolEntry
+{
     CAmount fee_individual;
-    CAmount fee_with_ancestors;
-    int64_t vsize_individual;
-    int64_t vsize_with_ancestors;
     const CTransaction& tx;
 
 public:
+    CAmount fee_with_ancestors;
+    int64_t vsize_individual;
+    int64_t vsize_with_ancestors;
     explicit MockMempoolEntry(CTxMemPool::txiter entry) :
         fee_individual{entry->GetModifiedFee()},
         fee_with_ancestors{entry->GetModFeesWithAncestors()},
```

glozow replied:
I'd say just remove the getters, it's fine to keep these public. This struct is only used by `MiniMiner`.
For background, I originally was trying to align the interface with `CTxMemPoolEntry` like `CTxMemPoolModifiedEntry` to reuse the `CompareTxMemPoolEntryByAncestorFee` comparator in a multi index container, but then realized using simple `std::map`s was enough.

t-bast commented on src/interfaces/chain.h:
```diff
@@ -205,6 +205,33 @@ class Chain
+    //  cost. If bump fees of descendants of the to-be-replaced transaction are requested, the value
+    //  will be 0. Fee-related RBF rules are not included as they are logically distinct.
+    //
+    //  Any outpoints that otherwise unavailable from the mempool (e.g. UTXOs from confirmed
```
nit:
```suggestion
    //  Any outpoints that are otherwise unavailable from the mempool (e.g. UTXOs from confirmed
```

t-bast commented on src/interfaces/chain.h:
```diff
@@ -205,6 +205,33 @@ class Chain
+    //  transactions or transactions not yet broadcast by the wallet) are given a bump fee of 0.
+    //
+    //  If multiple outpoints come from the same transaction (this should be very rare because
+    //  the transaction essentially multiple change outputs or paid the same wallet using multiple
```
nit:
```suggestion
    //  it means the transaction has multiple change outputs or paid the same wallet using multiple
```

t-bast reviewed (commented):
Concept ACK, thanks a lot for working on this, it will be very helpful!
And this set of internal utility functions will very likely be useful for many things in the future.
I'll create a set of E2E tests in eclair to run against this branch when I have a bit more time, I'll let you know the results.

t-bast commented:
I ran a first set of tests from within eclair against https://github.com/bitcoin/bitcoin/pull/26152/commits/a07ac02d2bc6500a03c29a0413bb913735dca46f, and everything is looking good :+1:

ishaanam reviewed (commented):
Concept ACK

Xekyo commented:
@jonatack, @t-bast: Thanks for the review and testing. I should have taken care of the nits. I made an attempt of getting rid of the getters on MockMempoolEntry, but what I did interfered with the calls made on properties of actual mempool entries. Will have to shift my approach.

Xekyo commented:
@glozow: Maybe for

> This needs to obey -maxtxfee

Maybe we can add a `maxtxfee` check to the filter introduced in #25729 for max weight after input sets are produced for different subsets of the available coins. Perhaps a separate PR that builds both on this one here and #25729.

Fixed whitespace issues, applied the propose change to a class for the `struct MockMempoolEntry`, amended comments in Chain interface.

glozow commented on src/wallet/coinselection.h:
```diff
@@ -66,7 +71,10 @@ struct COutput {
+    /** The fee necessary to bump this UTXO's ancestor transactions to the target feerate */
+    CAmount ancestor_bump_fees{0};
+
+    COutput(const COutPoint& outpoint, const CTxOut& txout, int depth, int input_bytes, bool spendable, bool solvable, bool safe, int64_t time, bool from_me, const std::optional<CFeeRate> feerate = std::nullopt, const std::optional<std::reference_wrapper<interfaces::Chain>> chain_interface = std::nullopt)
```
I have 2 concerns with this approach, please let me know what you think:

(1) The `COutput` constructor really shouldn't need to have a reference the chain interface; it smells a bit weird that coinselection (which I interpret to be a relatively well-modularized component thus far since it doesn't even depend on wallet) has a new dependency on interfaces/chain.h. It's unclear to me why `CalculateBumpFees()` needs to be called *inside* the constructor instead of just having `CAmount ancestor_bump_fees` be a parameter?

(2) This means `CalculateBumpFees()` will be called over and over again for each `COutput` constructed within the `AvailableCoins` loop. There's not a lot of duplicated work if all the outputs are from independent transactions, but it would definitely be faster to call it once with all the outpoints at once.

Approach-wise, I think it makes more sense to call `CalculateBumpFees()` just once, with the full list of outpoints. It should be fairly straightforward with preset inputs since you have the list already. For `AvailableCoins`, since they're constructed as you iterate through `mapWallet`, you could populate each output's `ancestor_bump_fees` values at the end. If you want to construct them as-is and then not mutate afterwards, then maybe do 2 passes for filtering and constructing?

Xekyo replied:
Now applying the bumpfees to each output after constructing the output.

glozow commented on test/functional/wallet_spend_unconfirmed.py:
```diff
@@ -0,0 +1,321 @@
+        self.test_chain_of_high_low_below_target_feerate()
+
+        # Actual test: Check that fee is calculated correctly when bumping while subtracting fee from output
+        self.test_target_feerate_unconfirmed_low_sffo()
```
in 337a24e6e4fd77d2fc3bf94e87c2a172c39a8fb0

Missing a test for `bumpfee` RPC?

Xekyo replied:
Added!

glozow reviewed (commented):
>> This needs to obey -maxtxfee

> Maybe we can add a maxtxfee check to the filter introduced in https://github.com/bitcoin/bitcoin/pull/25729 for max weight after input sets are produced for different subsets of the available coins. Perhaps a separate PR that builds both on this one here and https://github.com/bitcoin/bitcoin/pull/25729.

~Not sure if this is scope creep, but seems like breaking maxtxfee would be a bug and should probably be done in the same PR? Why not build this on top of #25729?~ ignore

Xekyo commented:
- Reordered members vs initialization
- Removed special casing of UTXOs without relatives in CalculateBumpFee
- Call CalculateBumpFee once for the whole UTXO pool instead of introducing chain-interface dependency on every UTXO

Todos:
- Prevent exceeding `maxtxfee`
- Add test for bumpfee RPC
- Add secondary modus for CalculateBumpFee that treats the provided UTXOs as being spent together. This allows us to recalculate the bumpfee of all inputs together to resolve our overpayment caveat.

Xekyo commented:
@glozow: looking more into this, I realized that the `maxtxfee` refers to an absolute fee, not a feerate. `maxtxfee` is checked after a transaction is built, so I don't see how ancestor aware funding changes anything in regard to `maxtxfee`—we still check at the end whether the amount of fee is allowed, regardless how we calculated the fee. If you meant `maxtxfeerate`, that is used to check raw transactions on submission in `sendrawtx`, so it doesn't apply here either.

achow101 commented:
> Question: is it better to only enforce `-maxtxfee` on the fees paid for the tx itself and not on the fees used to bump its ancestors? Or would the user expect that it's applied to any tx, bumping or not?

I think that `-maxtxfee` should be expected to behave the same regardless of bumping, it's a context-free check.

murchandamus committed:
Bump unconfirmed parent txs to target feerate

When a transaction uses an unconfirmed input, preceding this commit it
would not consider the feerate of the parent transaction. Given a parent
transaction with a lower ancestor feerate, this resulted in the new
transaction's ancestor feerate undershooting the target feerate.

This commit changes how we calculate the effective value of unconfirmed UTXOs.
The effective value of unconfirmed UTXOs is decreased by the fee
necessary to bump its ancestry to the target feerate. This also impacts
the calculation of the waste metric: since the estimate for the current
fee is increased by the bump fees, unconfirmed UTXOs current fees appear less
favorable compared to their unchanged long term fees.

This has one caveat: if multiple UTXOs have overlapping ancestries, each
of their individual estimates will account for bumping all ancestors.

Xekyo commented:
Latest changes:
- Fixed bug where the outpoints spent by a transaction to be replaced got set to a bumpfee of 0
- Added a test for preset inputs and the `bumpfee` RPC
- Rebased

Remaining Todo:
- Add secondary modus for CalculateBumpFee that treats the provided UTXOs as being spent together. This allows us to recalculate the bumpfee of all inputs together to resolve our overpayment caveat.

Xekyo commented:
Update:

<s>Caveat: If multiple UTXOs share ancestry, this implementation will overpay by bumping shared ancestors once per descendant.</s>

After calculating an input set candidate, we recalculate the bumpfee for the collective set of inputs and correct our fee estimation if there was an overestimate due to overlapping ancestries.

Left to do:
• clean up tests
• clean up commits
• touch up commit messages

Xekyo commented:
Cleaned up tests, redrew commits, touched up commit messages, rebased.

**Ready for review.** :partying_face:

ishaanam reviewed (commented):
Are there any functional tests that test for fee calculation for a transaction with both an input with an unconfirmed ancestor and a confirmed input? I couldn't find one by looking at the descriptions of the functional tests.

willcl-ark commented on test/functional/wallet_spend_unconfirmed.py:
```diff
@@ -0,0 +1,436 @@
+        self.assert_beats_target(ancestor_aware_tx)
+        resulting_ancestry_fee_rate = self.calc_set_fee_rate([parent_tx, ancestor_aware_tx])
+        assert_greater_than_or_equal(resulting_ancestry_fee_rate, self.target_fee_rate)
+        assert_greater_than_or_equal(self.target_fee_rate*1.5, resulting_ancestry_fee_rate)
```
Where does the 1.5 come from here (and in later tests e.g. `test_two_low_feerate_unconfirmed_parents()` 1.1)?

andrewtoth reviewed (commented):
Concept ACK

stickies-v reviewed (commented):
Concept ACK. Finished going through the code in the first 2 commits, more comments to come as I progress.

MarcoFalke commented:
Needs rebase after 8597260872bbef86524996ec695ffb30ec596416

LarryRuane commented on src/interfaces/chain.h:
```diff
@@ -213,6 +213,42 @@ class Chain
+    virtual std::map<COutPoint, CAmount> CalculateBumpFees(const std::vector<COutPoint>& outpoints, const CFeeRate& target_feerate) = 0;
+
+    //! Calculate the shared bump fees for a given set of outpoints per the
+    //  same strategy as in CalculateBumpFees(…).  Other than the above call,
```
```suggestion
    //  same strategy as in CalculateBumpFees(…). Unlike the above call,
```

LarryRuane commented on src/interfaces/chain.h:
```diff
@@ -213,6 +213,42 @@ class Chain
+    //  If the outpoint comes from an unconfirmed transaction that is already above the target
+    //  feerate or bumped by its descendant(s) already, it does not need to be bumped. Its bump fee
+    //  is 0. Likewise, if any of the transaction's ancestors are already bumped, they are not
+    //  included in the transaction's bump fee.
```
> they are not included

Is this true? IIUC, I don't think it works this way, or it shouldn't. If there are two outpoints sharing the same ancestor transaction, we don't know which of those two outpoints coin selection will choose if it chooses only one. If coin selection chooses the one for which we didn't bump to account for the ancestor, then our fee will be too low. I thought we bump both, and then _after_ coin selection, we make the adjustment.

LarryRuane commented on src/node/mini_miner.cpp:
```diff
@@ -0,0 +1,265 @@
+        } else {
+            auto outpoints_it = outpoints_needed_by_txid.find(txiter->GetTx().GetHash());
+            if (outpoints_it != outpoints_needed_by_txid.end()) {
+                for (const auto& outpoint : outpoints_it->second) {
+                    this->bump_fees.emplace(std::make_pair(outpoint, 0));
+                }
```
```suggestion
                for (const auto& outpoint : outpoints_it->second) {
                    bump_fees.emplace(outpoint, 0);
                }
                outpoints_needed_by_txid.erase(outpoints_it);
```
I'm not sure about this, but doing the `erase` here would be consistent with the `!mempool.exists()` case above (an entry for this transaction, none of whose outpoints we will use) is _not_ added to `outpoints_needed_by_txid`).

LarryRuane commented on src/node/mini_miner.cpp:
```diff
@@ -0,0 +1,265 @@
+                }
+            }
+        }
+        if (!remove) descendant_set_by_txid.emplace(std::make_pair(txid, cached_descendants));
```
```suggestion
        if (!remove) descendant_set_by_txid.emplace(std::make_pair(txid, std::move(cached_descendants)));
```

LarryRuane reviewed (commented):
898ad9d5904f1b689d18d94f20d92500cf443758
Concept ACK
LGTM, suggestions are minor. I'll continue reviewing.

LarryRuane commented on src/node/mini_miner.h:
```diff
@@ -0,0 +1,110 @@
+class MockMempoolEntry
+{
+    CAmount fee_individual;
+    const CTransaction& tx;
```
I'm not at all sure about this, but this is a reference variable so it's (sort of) a pointer, right? We continue to hold this reference after releasing `mempool.cs` -- is it possible for the transaction to go away (its memory deallocated) if, for example, it gets mined, thereby invalidating our reference? I wonder if this might be better:
```suggestion
    const CTransaction tx;
```
The debugger shows `sizeof(node::MockMempoolEntry)` is only 40 bytes with `tx` being a reference, and 152 if it's not a reference, which makes sense because `sizeof(CTransaction)` is 120.

Or, maybe it should be `const CTransactionRef tx;` because then it's a shared pointer, so if the tx gets removed from the mempool, our reference will remain valid until we're done with it. The advantage of this would be we wouldn't be making a full copy of the transaction.

LarryRuane commented on src/interfaces/chain.h:
```diff
@@ -213,6 +213,42 @@ class Chain
+    //  included in the transaction's bump fee.
+    //
+    //  This includes fee-bumping in RBFs. If an outpoint conflicts with another transaction in the
+    //  mempool, it is assumed that the goal is to replace that transaction. As such, the
+    //  calculation will exclude the to-be-replaced transaction, but will include the fee-bumping
+    //  cost. If bump fees of descendants of the to-be-replaced transaction are requested, the value
+    //  will be 0. Fee-related RBF rules are not included as they are logically distinct.
```
Would it be safer not to make this assumption? The cluster would be larger, but since the algorithms are efficient, the performance difference shouldn't be a problem. The `MiniMiner` constructor would be simpler too. Even if currently we're never given a outpoint that conflicts with a transaction _not_ being replaced, could that possibly change in the future?

Initially, I thought that was I'm suggesting here wouldn't work, because suppose that the transaction that our outpoint refers to (let's call it the parent) has a low feerate, but the transaction we're replacing (the parent's existing child) has a very high feerate. If, contrary to what the code currently does, we ignore the fact that we're replacing the child (we keep the child in the cluster, as if we're spending a different output of the parent), then we might conclude that we don't need to bump the parent's fee -- _whereas we actually do_, because the child will no longer exist!

But I think this won't happen because rule 6 of our [replacement policy](https://github.com/bitcoin/bitcoin/blob/master/doc/policy/mempool-replacements.md) requires the replacement transaction to have a higher feerate than all that it's replacing.

> 6. The replacement transaction's feerate is greater than the feerates of all directly conflicting transactions.

So it seems like even if we (mistakenly) think that the existing child will remain and that it will bump the parent somewhat, it won't bump it enough to meet our requested feerate, so we'll still bump it (as required by our desired feerate).

Xekyo replied:
I’m not sure I follow your train of thought here. I think you have since reviewed this further, do you think this still needs to be addressed?

Xekyo replied:
Thinking more about this, I think the assumption is valid and I don't expect it to change soon. It is also documented here, so if it changes, people should notice the conflict.

LarryRuane commented on src/node/mini_miner.cpp:
```diff
@@ -0,0 +1,265 @@
+        const auto ancestor_package_size = (*best_iter)->second.GetSizeWithAncestors();
+        const auto ancestor_package_fee = (*best_iter)->second.GetModFeesWithAncestors();
+        // Stop here. Everything that didn't "make it into the block" has bumpfee.
+        if (best_iter == entries.end() || ancestor_package_fee < target_feerate.GetFee(ancestor_package_size)) {
```
Simplification, we just asserted that it's not `entries.end()`
```suggestion
        if (ancestor_package_fee < target_feerate.GetFee(ancestor_package_size)) {
```

LarryRuane commented on src/node/mini_miner.cpp:
```diff
@@ -0,0 +1,265 @@
+        // Cache descendants for future use. Unlike the real mempool, a descendant MockMempoolEntry
+        // will not exist without its ancestor MockMempoolEntry, so these sets won't be invalidated.
+        std::vector<MockEntryMap::iterator> cached_descendants;
+        cached_descendants.emplace_back(entries_by_txid.find(txid));
```
898ad9d5904f1b689d18d94f20d92500cf443758 The result of this `find()` could be `entries_by_txid.end()` (not found), because in the previous loop (also over `cluster`), if the transaction is found in `to_be_replaced` (so we take the `else` path), then the transaction is not added to `entries_by_txid`. I think it turns out to be harmless, but I just wanted to point it out because it looks like it may be unintentional.

josibake replied:
Also noticed this and was a bit confused. In the case where `remove` is true, `cache_descendants` never gets added to the `descendants_by_txid` map, which is fine. In the case where `remove` is false, then `txid` should be in `entries_by_txid`, so it will get added to `cache_descendants`, along with all of `descendants`.

So what gets added to `descendants_by_txid` is `txidA: [txidA, txidB, txidC...]`, basically the parent + all of its children, which seems incorrect?

LarryRuane replied:
@josibake 
> ... which seems incorrect?

No, I think it's a convention in the strange Land of Mempool that a transaction's ancestor and descendant lists both include the transaction itself. This relates directly to my [comment](https://github.com/bitcoin/bitcoin/pull/26152#discussion_r1050404621) elsewhere that the `emplace_back()` should be removed; as it's currently written, the list would be: `txidA: [txidA, txidA, txidB, txidC...]`. This later causes a major accounting problem!

glozow replied:
Yes, it would be more correct to gate the `emplace_back()` on `!remove`.

LarryRuane commented on src/node/mini_miner.cpp:
```diff
@@ -0,0 +1,265 @@
+    assert(in_block.empty() || CFeeRate(total_fees, total_vsize) >= target_feerate);
+
+    // Each transaction that "made it into the block" has a bumpfee of 0, i.e. they are part of an
+    // ancestor package that exceeds the target feerate and don't need to be bumped.
```
```suggestion
    // ancestor package with at least the target feerate and don't need to be bumped.
```
(if the feerates are equal, no bump is required)

LarryRuane commented on src/node/mini_miner.h:
```diff
@@ -0,0 +1,110 @@
+class MiniMiner
+{
+    // Original outpoints requested
+    std::vector<COutPoint> requested_outpoints;
```
Should all these class variable names have the `m_` prefix?

Xekyo replied:
Renamed all MiniMiner class variables to be prefixed with `m_`

josibake reviewed (commented):
Concept ACK

Still in the process of reviewing (have reviewed up to and most of b669fd94f84e679d4549ef0abe1b0483e1406152), but left a few comments. 

Some general feedback: 

I noticed a lot of asserts being used. As I understand it, this is going to be a functionality exposed to the wallet, so wouldn't it be better to replace asserts with error messages that can be returned to the wallet? Seems better than crashing the node, IMO.

Also, the `MiniMiner` function is quite complicated. It might be worth adding a high-level overview to the PR description for the `MiniMiner::MiniMiner` as it seems like this is where a lot of the logic is. In particular, it seems to be working with several caches, so perhaps describing the general flow would be helpful. Take this feedback with a grain of salt, tho, as I'm not super familiar with the mempool and that could be why I'm struggling to follow the steps :sweat_smile:

glozow commented on src/node/mini_miner.cpp:
```diff
@@ -0,0 +1,220 @@
+    // Anything that's spent by the mempool is to-be-replaced
+    // Anything otherwise unavailable just has a bump fee of 0
+    for (const auto& outpoint : outpoints) {
+        if (const auto ptx{mempool.GetConflictTx(outpoint)}) {
+            // The conflicting transaction is to-be-replaced
+            to_be_replaced.insert(ptx->GetHash());
+        }
```
> To check my understanding:

> outpoints here refers to a set of unconfirmed UTXOs that our wallet would like to spend, meaning they are tx outputs. 

They are not necessarily unconfirmed UTXOs, just some specified outpoints/prevouts. They may refer to confirmed or unconfirmed UTXOs. They may refer to outputs that the node doesn't think exist.

> GetConflictTx checks if the outpoint exists in mapNextTx, which means it is the prevout (or input) into another tx in the mempool. If there is another transaction in the mempool spending one of these outputs, we put the tx which is spending the outpointt into to_be_replaced.

correct, `GetConflictTx` gives you another mempool transaction that spends the same tx.

> This is because we plan to construct a transaction which would then replace the conflicting tx, so we want to ignore doing anything with the conflicting txs for now?

We want to ensure we provide a bump fee for this UTXO, but ensure we calculate the bump fees *exluding the to-be-replaced transaction(s)*. Once they're replaced, they won't be there to bump their ancestors. So if we're replacing the child of a CPFP (e.g. to increase the bump even more), we want the correct fee to bump that parent *without* the original child there.

josibake replied:
Thanks, this explanation really helps!

> They are not necessarily unconfirmed UTXOs, just some specified outpoints/prevouts. They may refer to confirmed or unconfirmed UTXOs. They may refer to outputs that the node doesn't think exist.

Is it correct to say these `outpoints` are intended to be spent in a new transaction, as in the wallet has a set of UTXOs that it wants to spend in `TxB` where `TxB` can be a replacement of an already existing `TxA` or it can be an entirely new tx? From there, these `outpoints` can be a mix of confirmed/unconfirmed, but it's assumed at least one outpoint is unconfirmed

josibake replied:
Another thought I had while looking at this: we check if an `outpoint` has any conflicting spends in the mempool before we check if the outpoint itself is in the mempool. It seems we are inferring that the `outpoint` is in the mempool, otherwise it would not be present in `mapNextTx`. Is this the safest way to do this? Seems fine because `mapNextTx` is updated every time a tx is removed from the mempool, but figured I'd ask anyway in case there is a race condition / code path I'm not seeing which could make this unreliable in an edge case

glozow commented on src/txmempool.cpp:
```diff
@@ -1196,3 +1214,40 @@ const std::string RemovalReasonToString(const MemPoolRemovalReason& r) noexcept
+    std::vector<txiter> cluster{GetIterVec(txids)};
+    assert(cluster.size() == txids.size());
+    cluster.reserve(std::accumulate(cluster.cbegin(), cluster.cend(), 0, [](size_t sum, const auto it) {
+        return sum + it->GetCountWithAncestors() + it->GetCountWithDescendants(); }));
```
That's true, -1 makes sense. Though note this is an approximation rather than an exact reservation. It may overestimate because transactions may share ancestors/descendants, and may underestimate because the cluster may include more than just ancestors and descendants.

glozow commented on src/txmempool.cpp:
```diff
@@ -1196,3 +1214,40 @@ const std::string RemovalReasonToString(const MemPoolRemovalReason& r) noexcept
+        }
+        // i = index of where the list of unprocessed starts
+        for (size_t i{0}, unprocessed_count{txids.size()}; i < unprocessed_count; ++i) {
+            const auto curr = cluster[i];
+            for (const CTxMemPoolEntry& parent_entry : curr->GetMemPoolParentsConst()) {
+                const auto parent_it = mapTx.iterator_to(parent_entry);
+                if (!visited(parent_it)) {
+                    cluster.push_back(parent_it);
+                    // we still need to process this
+                    ++unprocessed_count;
+                }
+            }
+            for (const CTxMemPoolEntry& child_entry : curr->GetMemPoolChildrenConst()) {
+                const auto child_it = mapTx.iterator_to(child_entry);
+                if (!visited(child_it)) {
+                    cluster.push_back(child_it);
+                    // we still need to process this
+                    ++unprocessed_count;
+                }
+            }
```
Is this equivalent, given it's changing from `GetMemPoolChildrenConst` to `GetMemPoolChildren` and getting a mutable reference to `m_parents`?

glozow replied:
Mutable reference = the `merge` is modifying the entry's `m_parents` to now include its children as well. You can make a separate set that copies in the iterators from `GetMemPoolParentsConst` and `GetMemPoolChildrenConst`, but I'm not sure that's worth the lines of code reduction.

Xekyo replied:
Gonna skip this for now, unless there is more demand for it

glozow reviewed (commented):
Hopefully this is helpful and not annoying, but here's a branch that makes the first couple commits less assert-happy and applies some of the suggestions to miniminer: https://github.com/glozow/bitcoin/tree/26152-fixups

josibake commented:
> Hopefully this is helpful and not annoying, but here's a branch that makes the first couple commits less assert-happy and applies some of the suggestions to miniminer: https://github.com/glozow/bitcoin/tree/26152-fixups

lgtm!

josibake commented on src/node/mini_miner.cpp:
```diff
@@ -0,0 +1,220 @@
+        if (a_feerate != b_feerate) {
+            return a_feerate > b_feerate;
+        }
+        return &(*a) > &(*b);
```
what is the significance here? why not just have `return a_feerate >= b_feerate` or something like that?

Xekyo replied:
Added a comment

josibake commented:
Two thoughts:

1) I think this could be split up into two PRs; one for `MiniMiner` and another for the wallet logic. My reasoning here is the MiniMiner code seems like it could be useful outside of the wallet use case and thus could be merged independently and both parts by themselves seem complex enough that splitting it into two might make things easier to review.

2) per @LarryRuane 's comment [#26152 (comment)](https://github.com/bitcoin/bitcoin/pull/26152#discussion_r1038492630), I'm also not totally convinced we need the `to_be_replaced` logic in `MiniMiner`. I took out all of the code relating to `to_be_replaced` and the unit and functional tests still pass. It's possible this is due to insufficient test coverage, in which case having a test case to demonstrate why this logic is necessary would be super helpful. I have a few more thoughts regarding whether `MiniMiner` or the wallet should be handling the conflicting transactions logic, but I'll try to add a concrete example in a test case to first convince myself and also make the conversation more productive

glozow commented:
> I'm also not totally convinced we need the to_be_replaced logic in MiniMiner. I took out all of the code relating to to_be_replaced and the unit and functional tests still pass. It's possible this is due to insufficient test coverage

I would say this is at least 95% due to insufficient coverage in the unit tests (pretty sparse at the moment). I don't think it's possible to hit very much of the replacement-related logic through functional tests, because our (1) RBF rules currently do not allow any additional unconfirmed inputs so we won't actually ask MiniMiner to calculate much, (2) our wallet will never try to replace something with descendants and (3) our wallet will always use all the inputs from the replacee (so the size of `to_be_replaced` is always 1 and the set of unconfirmed ancestors is identical before and after).

But maybe this is a good reason *to* remove it! If it never executes because of how our wallet operates, and the wallet is the only client of this interface, then it would be complexity for no reason. We could just use `CalculateTotalBumpFees()` for a replacement (since that's all the unconfirmed inputs we'll use) and only add logic for replacements if (1) ever changes in the future.

josibake commented:
> But maybe this is a good reason _to_ remove it! If it never executes because of how our wallet operates, and the wallet is the only client of this interface, then it would be complexity for no reason.

I think this is a good reason to leave it out for now. I started writing a unit test to try and cover different replacement scenarios, but it felt a little silly to be writing tests for scenarios that would never actually happen to due the RBF rules and how the wallet behaves.

I also haven't been able to think of a scenario where `MiniMiner` would return an incorrect bumpfee if it didn't have the "to be replaced logic" if `MiniMiner` was given a set of "reasonable" outpoints in the first place. @Xekyo perhaps you have an example that I haven't thought of?

tldr; imo we should prefer a simpler implementation which makes as few assumptions as possible and add in more complexity if/when it is needed.

LarryRuane commented:
I had an intuition that the `MiniMiner` could be simpler, here's what I came up with, the latest commit on [this branch](https://github.com/LarryRuane/bitcoin/commits/2022-11-pr26152-refactor) (no PR), which is layered on top of this PR's commits. All the tests pass.

This is probably too large of a suggested change to this PR, given how many people have already reviewed it, so we should probably go ahead with this PR as is, and after it merges, I may offer this other implementation as a follow-up PR.

LarryRuane commented on src/node/mini_miner.cpp:
```diff
@@ -0,0 +1,265 @@
+            auto it = descendant_set_by_txid.find(anc->second.GetTx().GetHash());
+            assert(it != descendant_set_by_txid.end());
+            for (const auto& descendant : it->second) {
+                descendant->second.vsize_with_ancestors -= anc->second.GetTxSize();
+                descendant->second.fee_with_ancestors -= anc->second.GetModifiedFee();
```
You may want to assert that these don't go negative, I would actually do it before the decrement, like this:
```
assert(descendant->second.vsize_with_ancestors >= anc->second.GetTxSize());
descendant->second.vsize_with_ancestors -= anc->second.GetTxSize();
```

Xekyo replied:
Thanks, I’ve adopted your suggestion, although with an `Assume` instead.

LarryRuane commented on src/node/mini_miner.cpp:
```diff
@@ -0,0 +1,265 @@
+        // Cache descendants for future use. Unlike the real mempool, a descendant MockMempoolEntry
+        // will not exist without its ancestor MockMempoolEntry, so these sets won't be invalidated.
+        std::vector<MockEntryMap::iterator> cached_descendants;
+        cached_descendants.emplace_back(entries_by_txid.find(txid));
```
Something more important: this `emplace_back()` shouldn't be here at all, because each mempool entry is a member of its own descendants' list. So this "main" transaction gets added to `cached_descendants` below. This was causing the `fee_with_ancestors` and `vsize_with_ancestors` to go negative near the end of `BuildMockTemplate()` below.

LarryRuane commented on src/node/mini_miner.cpp:
```diff
@@ -0,0 +1,265 @@
+            in_block.insert(anc->second.GetTx().GetHash());
+            total_fees += anc->second.GetModifiedFee();
+            total_vsize += anc->second.GetTxSize();
+            auto it = descendant_set_by_txid.find(anc->second.GetTx().GetHash());
```
simpler:
```suggestion
            auto it = descendant_set_by_txid.find(anc->first);
```

LarryRuane commented on src/node/mini_miner.cpp:
```diff
@@ -0,0 +1,265 @@
+
+        // "Mine" all transactions in this ancestor set.
+        for (const auto& anc : ancestors) {
+            in_block.insert(anc->second.GetTx().GetHash());
```
```suggestion
            in_block.insert(anc->first);
```
(functionally the same, just simpler)

LarryRuane reviewed (commented):
Just FYI, I added a commit with a benchmark to the [branch](https://github.com/LarryRuane/bitcoin/commits/2022-11-pr26152-refactor) I mentioned earlier (with an alternate implementation of `MiniMiner`); feel free to cherry-pick this commit. It also seems useful as a stress test -- most of my comments below are the result of running the benchmark.

Xekyo commented on src/txmempool.cpp:
```diff
@@ -952,6 +953,24 @@ CTxMemPool::setEntries CTxMemPool::GetIterSet(const std::set<uint256>& hashes) c
+    ret.reserve(txids.size());
+    for (const auto& txid : txids) {
+        const auto it = GetIter(txid);
+        assert(it);
```
Fixed

Xekyo commented on src/txmempool.cpp:
```diff
@@ -1196,3 +1214,40 @@ const std::string RemovalReasonToString(const MemPoolRemovalReason& r) noexcept
+{
+    AssertLockHeld(cs);
+    std::vector<txiter> cluster{GetIterVec(txids)};
+    assert(cluster.size() == txids.size());
```
Fixed

Xekyo commented on src/txmempool.h:
```diff
@@ -646,9 +646,16 @@ class CTxMemPool
+     * only returns the ones that exist. */
     setEntries GetIterSet(const std::set<uint256>& hashes) const EXCLUSIVE_LOCKS_REQUIRED(cs);
 
+    /** Translate a list of hashes into a list of mempool iterators to avoid repeated lookups.
+     * The nth element in txids becomes the nth element in the returned vector. If any of the txids
+     * don't actually exist in the mempool, returns an empty vector. */
+    std::vector<txiter> GetIterVec(const std::vector<uint256>& txids) const EXCLUSIVE_LOCKS_REQUIRED(cs);
+
```
Fixed

murchandamus reviewed (commented):
WIP: I’ve addressed all comments up to https://github.com/bitcoin/bitcoin/pull/26152#issuecomment-1344506798

I am currently evaluating whether splitting this PR in two makes sense, and still need to investigate @LarryRuane’s alternative branch.

Redesignating this PR to draft to signal that it’s not ready for review at the moment.

Xekyo commented on src/node/mini_miner.cpp:
```diff
@@ -0,0 +1,265 @@
+            const CTransaction& tx = (*iter)->second.GetTx();
+            for (const auto& input : tx.vin) {
+                if (auto parent_it{entries_by_txid.find(input.prevout.hash)}; parent_it != entries_by_txid.end()) {
+                    to_process.insert(parent_it);
+                    ancestors.insert(parent_it);
```
Great, thanks

Xekyo commented on src/node/mini_miner.h:
```diff
@@ -0,0 +1,110 @@
+class MiniMiner
+{
+    // Original outpoints requested
+    std::vector<COutPoint> requested_outpoints;
```
Removing `requested_outpoints` in favor of the keys of `bump_fees` breaks tests, so I assume they’re not overlapping exactly.

Xekyo commented on src/txmempool.cpp:
```diff
@@ -1196,3 +1214,40 @@ const std::string RemovalReasonToString(const MemPoolRemovalReason& r) noexcept
+    AssertLockHeld(cs);
+    std::vector<txiter> cluster{GetIterVec(txids)};
+    assert(cluster.size() == txids.size());
+    cluster.reserve(std::accumulate(cluster.cbegin(), cluster.cend(), 0, [](size_t sum, const auto it) {
```
We don't know the cluster size before, so this doesn’t allow us to optimize here

LarryRuane commented:
For anyone wanting to review this PR and would like some help with basic mempool concepts, I made a video: https://youtu.be/sQ05azzTp9o.

Xekyo commented:
Rebased on latest version of #27021

Xekyo commented:
Rebased on latest version of https://github.com/bitcoin/bitcoin/pull/27021. If you are interested in having #26152 in Bitcoin Core v25.0, please consider making time to review #27021.

Xekyo commented:
Rebased on latest version of #27021, then rebased on master to resolve merge conflicts. 

Needed to reintroduce access to the Chain interface in `ChooseSelectionResult()`.

Xekyo commented:
Rebased on #27021, need to rebase on master next, then will incorporate the follow-up nits from #27021

murchandamus committed:
Match tx names to index in miniminer overlap test

Follow-up from #27021: In the prior commit, the vector started counting
at 0, but the transaction names started with 1. This commit matches the
names to the transactions’ vector indices for better readability.

Co-authored-by: theStack <sebastian.falbesoner@gmail.com>

murchandamus committed:
Remove unused imports

Follow-up from #27021

murchandamus committed:
Make MiniMinerMempoolEntry fields private

Follow-up from #27021: accessing of fields in MiniMinerMempoolEntry was
done inconsistently. Even though we had a getter, we would directly
write to the fields when we needed to update them.
This commits sets the fields to private and introduces a method for
updating the ancestor information in transactions using the same method
name as used for Mempool Entries.

Xekyo commented:
Rebased on master, incorporated outstanding follow-ups from https://github.com/bitcoin/bitcoin/pull/27021#issuecomment-1554675156

murchandamus committed:
Fix calculation of ancestor set feerates in test

Follow-up from #27021.
Also included is an ASCII art visualization of the test’s transaction
topology by theStack.

Co-authored-by: theStack <sebastian.falbesoner@gmail.com>

Xekyo commented:
Rebased, added four commits for the follow-ups from #27021, cleaned up the commit messages, added @theStack’s wonderful topology overview for the transactions, built each commit separately to make sure all tests pass.

**Ready for review**

S3RK commented on src/wallet/spend.cpp:
```diff
@@ -1008,7 +1025,9 @@ static util::Result<CreatedTransactionResult> CreateTransactionInternal(
     // and in the spirit of "smallest possible change from prior
     // behavior."
     const uint32_t nSequence{coin_control.m_signal_bip125_rbf.value_or(wallet.m_signal_rbf) ? MAX_BIP125_RBF_SEQUENCE : CTxIn::MAX_SEQUENCE_NONFINAL};
+    CAmount total_bump_fees{0};
```
in "Bump unconfirmed parent txs to target feerate" 8e3d924f02ee28043e8844a301389915823e5893

nit: we can encapsulate bump fee calculation within `CoinSelectionResult::GetTotalBumpFee()`. It will come handy in the next commit as well

Xekyo replied:
I tried this change, and remembered why I put in the "discount" in the first place: I think I have incorporated the bump fees in the effective values before, so I need to know whether there is a difference between the sum of the bump fees and the combined inputs’ bump fee. I have an idea how to incorporate your suggestion, but I gotta try tomorrow.

S3RK reviewed (commented):
Did a first pass of code review. Overall code looks really good, need to spend more time reviewing the tests though.

murchandamus reviewed (commented):
Took one suggestion, will revisit the other tomorrow.

Xekyo commented on src/wallet/coinselection.cpp:
```diff
@@ -512,7 +520,12 @@ CAmount SelectionResult::GetSelectedValue() const
+    return std::accumulate(m_selected_inputs.cbegin(), m_selected_inputs.cend(), CAmount{0}, [](CAmount sum, const auto& coin) { return sum + coin->GetEffectiveValue(); }) + bump_fee_group_discount;
+}
+
+CAmount SelectionResult::GetBumpFeeDiscount() const
```
I added a function `GetSummedBumpFees()`, but I have yet to find a neat way how I can drop the discount. I was experimenting with having both a function `GetSummedBumpFees()` and a function `GetCombinedBumpFee()` on SelectionResult, but it made it more complicated because I now had to get the Chain interface and target feerate into `SelectionResult`. I am tending to leaving it as it is now, unless I come up with a more elegant way to transition.

S3RK replied:
My idea was to add `bump_fee_group_discount` within `GetTotalBumpFee()` function. Probably I'm missing something but I don't know what. Why wouldn't it just work?

Xekyo replied:
I would prefer that the method `GetTotalBumpFee()` does not either return the sum of individual bump fees or the combined bump fee for the input set depending on when it is called. I’d be worried that it would be hard to understand later and might be a source of future bugs.

Xekyo commented on src/wallet/spend.cpp:
```diff
@@ -611,6 +611,28 @@ util::Result<SelectionResult> ChooseSelectionResult(const CAmount& nTargetValue,
+    // If the chosen input set has unconfirmed inputs, check for synergies from overlapping ancestry
+    for (auto& result : results) {
+        std::vector<COutPoint> outpoints;
+        CAmount summed_bump_fees = 0;
```
Added `GetSummedBumpFees()`

Xekyo commented on src/wallet/spend.cpp:
```diff
@@ -573,6 +573,28 @@ util::Result<SelectionResult> ChooseSelectionResult(const CWallet& wallet, const
+        if (bump_fee_overestimate) {
+            result.SetBumpFeeDiscount(bump_fee_overestimate);
+            // Update waste
+            result.ComputeAndSetWaste(coin_selection_params.min_viable_change, coin_selection_params.m_cost_of_change, coin_selection_params.m_change_fee);
```
Still open

Xekyo replied:
Did about 40M executions, and no issues! Gonna calculate the Waste score below here as discussed.

S3RK replied:
In theory the waste score should be the same.

Because not all of the params are available within BnB we call
`result.ComputeAndSetWaste(cost_of_change, cost_of_change, 0);`
instead of
`result.ComputeAndSetWaste(min_viable_change, cost_of_change, change_fee);`

I believe this should be fine as BnB will not produce results with more than `cost_of_change` excess and therefore `GetChange()` will always return 0 and then `change_fee` doesn't matter.

murchandamus reviewed (commented):
Renamed the functions on the Chain interface to better express their utility. Added a convenience function for accessing bump fees to SelectionResult.

achow101 commented on src/wallet/spend.cpp:
```diff
@@ -346,6 +351,15 @@ CoinsResult AvailableCoins(const CWallet& wallet,
+    if (feerate.has_value()) {
+        std::map<COutPoint, CAmount> map_of_bump_fees = wallet.chain().CalculateIndividualBumpFees(outpoints, feerate.value());
+
+        for (auto& output : result.All()) {
+            output.ApplyBumpFee(map_of_bump_fees.at(output.outpoint));
```
In f2daebf8d1ac3829faa6443425cb97a763e9f780 "Bump unconfirmed parent txs to target feerate"

This doesn't actually apply the bumpfees to all of the outputs in `result. `All()` will return a copy of all of the outputs, not references to them, so this is just applying the bumpfees to a temporary.

Because of this, the tests added in this commit fail. They only do not fail because of the adjustment to the fee that is done in the following commit.

achow101 commented on test/functional/wallet_spend_unconfirmed.py:
```diff
@@ -0,0 +1,460 @@
+        self.def_wallet  = self.nodes[0].get_wallet_rpc(self.default_wallet_name)
+        self.generate(self.nodes[0], 110)
+
+        # Test that assumptions about meeting feerate and being able to test it hold
```
In f2daebf8d1ac3829faa6443425cb97a763e9f780 "Bump unconfirmed parent txs to target feerate"

nit: I prefer that comments about what a test does are attached to the test implementation itself rather than the caller.

achow101 commented on src/wallet/feebumper.cpp:
```diff
@@ -80,7 +80,19 @@ static feebumper::Result CheckFeeRate(const CWallet& wallet, const CFeeRate& new
+        reused_inputs.push_back(txin.prevout);
+    }
+
+    std::map<COutPoint, CAmount> bump_fees = wallet.chain().CalculateBumpFees(reused_inputs, newFeerate);
```
In f2daebf8d1ac3829faa6443425cb97a763e9f780 "Bump unconfirmed parent txs to target feerate"

Compilation failure in this commit.

```
wallet/feebumper.cpp: In function ‘wallet::feebumper::Result wallet::CheckFeeRate(const CWallet&, const CMutableTransaction&, const CFeeRate&, int64_t, CAmount, std::vector<bilingual_str>&)’:
wallet/feebumper.cpp:89:61: error: ‘class interfaces::Chain’ has no member named ‘CalculateBumpFees’
   89 |     std::map<COutPoint, CAmount> bump_fees = wallet.chain().CalculateBumpFees(reused_inputs, newFeerate);
      |                                                             ^~~~~~~~~~~~~~~~~
```

This function was renamed to `CalculateIndividualBumpFees`

achow101 commented on src/wallet/spend.cpp:
```diff
@@ -346,6 +351,18 @@ CoinsResult AvailableCoins(const CWallet& wallet,
+    if (feerate.has_value()) {
+        std::map<COutPoint, CAmount> map_of_bump_fees = wallet.chain().CalculateBumpFees(outpoints, feerate.value());
+
+        for (auto& [_, outputs] : result.coins) {
```
This is incorrect, see https://github.com/bitcoin/bitcoin/pull/26152#discussion_r1248205239

S3RK replied:
Oops, sorry about misleading you 🙄

Xekyo replied:
Ah drats, I did not realize that `CoinsResult.All()` produces a copy. Reverting this.

S3RK commented on test/functional/wallet_spend_unconfirmed.py:
```diff
@@ -0,0 +1,460 @@
+        parent_tx = wallet.gettransaction(txid=parent_txid, verbose=True)
+
+        self.assert_undershoots_target(parent_tx)
+        resulting_fee_rate_funding = self.calc_fee_rate(parent_tx)
```
that seems like a duplicate check of `assert_undershoots_target`

S3RK reviewed (commented):
Started reviewing tests, but need more time.

t-bast reviewed (commented):
I tried to run my own set of external e2e tests against https://github.com/bitcoin/bitcoin/pull/26152/commits/43b86e49351b3160c2e39807997a9373cd88e173 and they all hit the following assertion error:

```
bitcoind: wallet/coinselection.cpp:495: void wallet::SelectionResult::SetBumpFeeDiscount(CAmount): Assertion `discount >= 0' failed
```

Reading through recent comments, it seems somewhat expected that this version isn't yet fully working, let me know when you want me to run my tests again!

Xekyo commented on src/wallet/spend.h:
```diff
@@ -148,8 +150,7 @@ util::Result<SelectionResult> AttemptSelection(const CAmount& nTargetValue, Outp
  */
-util::Result<SelectionResult> ChooseSelectionResult(const CAmount& nTargetValue, Groups& groups, const CoinSelectionParams& coin_selection_params);
-
+util::Result<SelectionResult> ChooseSelectionResult(const CWallet& wallet, const CAmount& nTargetValue, Groups& groups, const CoinSelectionParams& coin_selection_params);
```
Done

murchandamus reviewed (commented):
Addressed comments by @S3RK and @achow101, and rebased to fix a conflict.

Xekyo commented:
> Reading through recent comments, it seems somewhat expected that this version isn't yet fully working, let me know when you want me to run my tests again!

Sorry about that, I was a bit in a rush on Friday. I’ve fixed a number of issues, if this latest push’s checks turn up green, it might be more fruitful to try now.

t-bast commented:
> I’ve fixed a number of issues, if this latest push’s checks turn up green, it might be more fruitful to try now.

I confirm that, I've ran my set of tests against [eclair](https://github.com/ACINQ/eclair) and everything looks good using https://github.com/bitcoin/bitcoin/pull/26152/commits/0f6c13665c4ab7d4928ff0fa63c4e755667f7fd6 :tada: 

The tests I ran can be found in the last two commits of [this branch](https://github.com/ACINQ/eclair/commits/wip-test-ancestor-aware-funding-bitcoind-25).

achow101 commented on src/wallet/feebumper.cpp:
```diff
@@ -86,37 +86,35 @@ static feebumper::Result CheckFeeRate(const CWallet& wallet, const CMutableTrans
     }
-
-    CAmount new_total_fee = newFeerate.GetFee(maxTxSize) + total_bump_fees;
+    CAmount fees_including_bump_fees = newFeerate.GetFee(maxTxSize) + combined_bump_fee.value();
```
In 0f6c13665c4ab7d4928ff0fa63c4e755667f7fd6 "Amend bumpfee for inputs with overlapping ancestry"

nit: I'd prefer if variables were not renamed if they don't need to be, it makes review of this commit slightly harder and unnecessarily breaks git blame. `new_total_fee` is still a valid description of what this variable contains.

achow101 commented on src/wallet/spend.cpp:
```diff
@@ -1034,7 +1052,7 @@ static util::Result<CreatedTransactionResult> CreateTransactionInternal(
         return util::Error{_("Missing solving data for estimating transaction size")};
     }
-    CAmount fee_needed = coin_selection_params.m_effective_feerate.GetFee(nBytes) + result.GetSummedBumpFees();
+    CAmount fee_needed = coin_selection_params.m_effective_feerate.GetFee(nBytes) + result.GetSummedBumpFees() - result.GetBumpFeeDiscount();
```
In 0f6c13665c4ab7d4928ff0fa63c4e755667f7fd6 "Amend bumpfee for inputs with overlapping ancestry"

`GetBumpFeeDiscount` is only used here. It doesn't seem like it's useful to require the caller to calculate the final bumpfee rather than having `SelectionResult` just do that when the the discount is applied and have a single method to return the bumpfees required for a particular selection.

Xekyo replied:
I’ve been doing some staring at this today. It seems to me that the waste score of finished input sets should also be adjusted according to the bump fee discount. Gotta do more pondering, will get back to this.

S3RK replied:
Yes, the discount value is a function of the effective fee rate, therefore it has to be taken into account in waste calculation. The problem is that GetSelectionWaste doesn't use SelectionResult.GetEffectiveValue() which already accounts for bump fee discoun

murchandamus replied:
Okay, I replaced `GetSummedBumpFees()` with `GetTotalBumpFees()` which deducts the `bump_fee_discount` if it has been set.

achow101 commented on src/wallet/spend.cpp:
```diff
@@ -609,6 +607,26 @@ util::Result<SelectionResult> ChooseSelectionResult(const CAmount& nTargetValue,
+        CAmount bump_fee_overestimate = result.GetSummedBumpFees() - combined_bump_fee.value();
+        if (bump_fee_overestimate) {
+            result.SetBumpFeeDiscount(bump_fee_overestimate);
+            // Update waste
```
In 0f6c13665c4ab7d4928ff0fa63c4e755667f7fd6 "Amend bumpfee for inputs with overlapping ancestry"

nit: this comment seems to be misplaced?

S3RK replied:
Yes, the discount value is a function of the effective fee rate, therefore it has to be taken into account in waste calculation. The problem is that `GetSelectionWaste` doesn't use `SelectionResult.GetEffectiveValue()` which already accounts for bump fee discount.

S3RK commented on test/functional/wallet_spend_unconfirmed.py:
```diff
@@ -0,0 +1,458 @@
+        vsizes = sum([tx["decoded"]["vsize"] for tx in txs])
+        return fees / vsizes
+
+    def assert_spends_only_parent(self, tx, parent_txid):
```
This could be extended to a set of parent txs to verify cases with multiple parents

murchandamus replied:
Thanks, good idea. I went with:

```python
    def assert_spends_only_parents(self, tx, parent_txids):
        parent_checklist = parent_txids.copy()
        number_inputs = len(tx["decoded"]["vin"])
        assert_equal(number_inputs, len(parent_txids))
        for i in range(number_inputs):
            txid_of_input = tx["decoded"]["vin"][i]["txid"]
            assert txid_of_input in parent_checklist
            parent_checklist.remove(txid_of_input)
```

S3RK commented on test/functional/wallet_spend_unconfirmed.py:
```diff
@@ -0,0 +1,458 @@
+        assert_greater_than_or_equal(resulting_ancestry_fee_rate, self.target_fee_rate)
+        assert_greater_than_or_equal(self.target_fee_rate*1.1, resulting_ancestry_fee_rate)
+        resulting_ancestry_fee_rate_with_high_feerate_gp = self.calc_set_fee_rate([gp_tx, p_tx, ancestor_aware_tx])
+        assert_greater_than_or_equal(resulting_ancestry_fee_rate_with_high_feerate_gp, self.target_fee_rate*1.1)
```
nit: why there is a multiplier here?

murchandamus replied:
The grandparent transaction has a higher feerate, so we only need to bump the parent. We check that if we get the complete chain of unconfirmed transactions, that the resulting feerate is above our target.

S3RK commented on src/wallet/spend.cpp:
```diff
@@ -195,6 +196,7 @@ util::Result<PreSelectedInputs> FetchSelectedInputs(const CWallet& wallet, const
 
         /* Set some defaults for depth, spendable, solvable, safe, time, and from_me as these don't matter for preset inputs since no selection is being done. */
         COutput output(outpoint, txout, /*depth=*/ 0, input_bytes, /*spendable=*/ true, /*solvable=*/ true, /*safe=*/ true, /*time=*/ 0, /*from_me=*/ false, coin_selection_params.m_effective_feerate);
+        output.ApplyBumpFee(map_of_bump_fees.at(output.outpoint));
```
Isn't `test_preset_input_cpfp` testing this?

ishaanam replied:
In that test don't all of the pre-selected inputs belong to the wallet?

S3RK replied:
ah, yes, you're right

S3RK commented on test/functional/wallet_spend_unconfirmed.py:
```diff
@@ -0,0 +1,458 @@
+
+        self.assert_undershoots_target(parent_tx)
+
+        to_be_rbfed_ancestor_aware_txid = wallet.sendtoaddress(address=self.def_wallet.getnewaddress(), amount=0.5, fee_rate=self.target_fee_rate, subtractfeefromamount=True)
```
nit: sffo is not needed

murchandamus replied:
Removed SSFO from this test

S3RK reviewed (commented):
Continue tests review

ismaelsadeeq reviewed (commented):
Light code review ACK for commit 0f6c13665c4ab7d4928ff0fa63c4e755667f7fd6:

As I continue familiarizing myself with the code, my understanding is that this PR addresses follow-up nits up to a3b178f3c082453120e816945cea220bc7397ee5. Additionally, it introduces the use of `MiniMiner` to calculate the necessary fees required for fee bumping unconfirmed inputs to a target fee rate. The calculated fee bumps values are then added to the transaction's fee to account for the ancestor fee bump and reach the target fee rate

Before this PR, unconfirmed transaction outputs were used as inputs in a new transaction without considering the ancestor fee bump. This could result in the transaction having a lower ancestor fee rate if the unconfirmed ancestor's fee rate is low.  with this PR, we can now ensure that the necessary fees to bump a low-fee ancestor to the target fee rate, if required, are added to the transaction's fee rate.

This looks good to me, I just suggest nits and questions

murchandamus commented on src/wallet/coinselection.cpp:
```diff
@@ -7,6 +7,7 @@
 #include <common/system.h>
 #include <consensus/amount.h>
 #include <consensus/consensus.h>
+#include <interfaces/chain.h>
```
Ah thanks, that belongs to the commit one later, where we need the chain interface to amend the bump fees in case of overlapping ancestries.

achow101 committed:
coinselection: Move GetSelectionWaste into SelectionResult

GetSelectionWaste will need to access more context within a selection
result, and so should be a private member function rather than a static
function. It's only use outside of SelectionResult was for tests which
have now been updated to just make a SelectionResult.

Co-authored-by: Murch <murch@murch.one>

murchandamus commented on test/functional/wallet_spend_unconfirmed.py:
```diff
@@ -0,0 +1,458 @@
+        ancestor_aware_tx = wallet.gettransaction(txid=ancestor_aware_txid, verbose=True)
+        self.assert_beats_target(ancestor_aware_tx)
+
+        self.generate(self.nodes[0], 1)
```
That’s right, I’ll remove the unnecessary block generation

murchandamus commented on test/functional/wallet_spend_unconfirmed.py:
```diff
@@ -0,0 +1,458 @@
+        self.assert_beats_target(ancestor_aware_tx)
+        resulting_ancestry_fee_rate = self.calc_set_fee_rate([gp_tx, p_tx, ancestor_aware_tx])
+        assert_greater_than_or_equal(resulting_ancestry_fee_rate, self.target_fee_rate)
+        assert_greater_than_or_equal(self.target_fee_rate*1.5, resulting_ancestry_fee_rate)
```
Every once in a while we’ll get a signature that’s gonna be one byte smaller, so a little bit of a buffer will prevent intermittent failures. I reduced the buffers to 1%, though.

murchandamus commented on src/test/miniminer_tests.cpp:
```diff
@@ -280,141 +280,141 @@ BOOST_FIXTURE_TEST_CASE(miniminer_overlap, TestChain100Setup)
     LOCK2(::cs_main, pool.cs);
     TestMemPoolEntryHelper entry;
 
```
I also updated the `miniminer_1p1c` test to match the transaction numbers to the indices

murchandamus commented on src/node/mini_miner.h:
```diff
@@ -38,6 +39,10 @@ class MiniMinerMempoolEntry
     int64_t GetTxSize() const { return vsize_individual; }
     int64_t GetSizeWithAncestors() const { return vsize_with_ancestors; }
     const CTransaction& GetTx() const LIFETIMEBOUND { return *tx; }
+    void UpdateAncestorState(int64_t vsize_change, CAmount fee_change) {
```
That’s a good idea, but I’m gonna leave this, because this matches how we interact with the mempool entries.

murchandamus commented on src/test/miniminer_tests.cpp:
```diff
@@ -327,10 +354,16 @@ BOOST_FIXTURE_TEST_CASE(miniminer_overlap, TestChain100Setup)
-    const auto tx7_anc_feerate = CFeeRate(low_fee + high_fee, tx_vsizes[5] + tx_vsizes[7]);
+    const auto tx6_anc_feerate = CFeeRate(high_fee + low_fee + med_fee, tx_vsizes[4] + tx_vsizes[5] + tx_vsizes[6]);
+    const auto tx6_iter = pool.GetIter(tx6->GetHash());
+    BOOST_CHECK(tx6_anc_feerate == CFeeRate(tx6_iter.value()->GetModFeesWithAncestors(), tx6_iter.value()->GetSizeWithAncestors()));
```
Yes, we are doing it manually, because we also compare it against what the mempool thinks it is.

murchandamus commented on src/wallet/coinselection.h:
```diff
@@ -104,6 +109,15 @@ struct COutput {
+    {
+        assert(bump_fee >= 0);
+        ancestor_bump_fees = bump_fee;
+        *fee += bump_fee;
```
I added an `assert(fee);` right before.

murchandamus commented:
Sorry, I still need to squash this, but this should be addressing all the open comments

murchandamus commented:
I believe that I have now addressed all open feedback. Ready for review

murchandamus commented on test/functional/wallet_spend_unconfirmed.py:
```diff
@@ -0,0 +1,458 @@
+        parent_txid = wallet.sendtoaddress(address=wallet.getnewaddress(), amount=1.5, fee_rate=2)
+        p_tx = wallet.gettransaction(txid=parent_txid, verbose=True)
+
+        self.assert_undershoots_target(p_tx)
```
Added

murchandamus commented:
Rebased to get rid of CI issues.

S3RK commented on src/wallet/coinselection.cpp:
```diff
@@ -488,14 +491,22 @@ CAmount GenerateChangeTarget(const CAmount payment_value, const CAmount change_f
 {
     const CAmount change = GetChange(min_viable_change, change_fee);
 
     if (change > 0) {
```
nit: this `if` can be removed and instead we can check if change exist directly in `GetChange()`

murchandamus replied:
I started with that, and then realized that `ComputeAndSetWaste` and `GetSelectionWaste` now live in the same class, so the former essentially wrapping the latter doesn’t make sense anymore. Then I cleaned that up, but that broke a bunch of the waste calculation tests. So, finally, I propose to clean all this up in a follow-up PR, which I will open shortly.

S3RK replied:
sure, that's definitely cosmetics and can wait for a follow up 👍

murchandamus replied:
I’ve opened https://github.com/bitcoin/bitcoin/pull/28366 to collect these changes

S3RK commented on src/wallet/coinselection.cpp:
```diff
@@ -449,19 +450,21 @@ void OutputGroupTypeMap::Push(const OutputGroup& group, OutputType type, bool in
 }
 
-CAmount GetSelectionWaste(const std::set<std::shared_ptr<COutput>>& inputs, CAmount change_cost, CAmount target, bool use_effective_value)
+CAmount SelectionResult::GetSelectionWaste(CAmount change_cost, CAmount target, bool use_effective_value)
```
we need tests for waste calculation with bumpfee and especially for bumpfee discount

murchandamus replied:
Added a test for waste calculation with bumpfee and bumpfee discount.

S3RK commented on src/wallet/coinselection.cpp:
```diff
@@ -449,19 +450,21 @@ void OutputGroupTypeMap::Push(const OutputGroup& group, OutputType type, bool in
-        selected_effective_value += use_effective_value ? coin.GetEffectiveValue() : coin.txout.nValue;
     }
+    // Bump fee of whole selection may diverge from sum of individual bump fees
+    waste += GetTotalBumpFees();
```
You only need to detract the discount, because `coin.GetFee()` already accounts for initial bump

S3RK commented on src/wallet/coinselection.cpp:
```diff
@@ -449,19 +450,21 @@ void OutputGroupTypeMap::Push(const OutputGroup& group, OutputType type, bool in
     }
+    // Bump fee of whole selection may diverge from sum of individual bump fees
+    waste += GetTotalBumpFees();
+    selected_effective_value += use_effective_value ? GetSelectedEffectiveValue() : GetSelectedValue();
```
nit: move the `selected_effective_value` var and the operation under the `else` below

murchandamus reviewed (commented):
Addressed S3RK’s comments, added waste calculation test with bump fees and group discount

murchandamus commented:
Rebased. (Ready for review.)

t-bast reviewed (approved):
ACK https://github.com/bitcoin/bitcoin/pull/26152/commits/1b2fb54d5c18579e013b1dfd061c2b29d8cffdc2, I ran my e2e lightning tests on top of that branch and everything is looking good :+1:

ismaelsadeeq commented:
re ACK 1b2fb54d5c18579e013b1dfd061c2b29d8cffdc2

Reviewed up to acb6aaeac49dd4543280653f19d2377ffb175c9e and functional test
As I am not really familiar with the coinselection part of the PR, I lightly review the commits.
I rebased latest master a4e0bcb6c9a8db5a74c74c5cddbb065ba9182482 on this PR, the tests failed and passed after making the branch.

achow101 commented:
ACK 1b2fb54d5c18579e013b1dfd061c2b29d8cffdc2

brunoerg commented on src/wallet/spend.cpp:
```diff
@@ -162,6 +162,7 @@ util::Result<PreSelectedInputs> FetchSelectedInputs(const CWallet& wallet, const
 {
     PreSelectedInputs result;
     const bool can_grind_r = wallet.CanGrindR();
+    std::map<COutPoint, CAmount> map_of_bump_fees = wallet.chain().CalculateIndividualBumpFees(coin_control.ListSelected(), coin_selection_params.m_effective_feerate);
```
non blocker, just a suggestion: In 35a35fa989f14406f8218c20c4e4e732cee0bd6c: you could move `map_of_bump_fees` to right before `ApplyBumpFee`, this way we calculate the individual bump fees when we really sure we're going to use it: 
```diff
diff --git a/src/wallet/spend.cpp b/src/wallet/spend.cpp
index 15004c67b0..db0870f31b 100644
--- a/src/wallet/spend.cpp
+++ b/src/wallet/spend.cpp
@@ -162,8 +162,8 @@ util::Result<PreSelectedInputs> FetchSelectedInputs(const CWallet& wallet, const
 {
     PreSelectedInputs result;
     const bool can_grind_r = wallet.CanGrindR();
-    std::map<COutPoint, CAmount> map_of_bump_fees = wallet.chain().CalculateIndividualBumpFees(coin_control.ListSelected(), coin_selection_params.m_effective_feerate);
-    for (const COutPoint& outpoint : coin_control.ListSelected()) {
+    const auto list_selected{coin_control.ListSelected()};
+    for (const COutPoint& outpoint : list_selected) {
         int input_bytes = -1;
         CTxOut txout;
         if (auto ptr_wtx = wallet.GetWalletTx(outpoint.hash)) {
@@ -198,6 +198,7 @@ util::Result<PreSelectedInputs> FetchSelectedInputs(const CWallet& wallet, const
 
         /* Set some defaults for depth, spendable, solvable, safe, time, and from_me as these don't matter for preset inputs since no selection is being done. */
         COutput output(outpoint, txout, /*depth=*/ 0, input_bytes, /*spendable=*/ true, /*solvable=*/ true, /*safe=*/ true, /*time=*/ 0, /*from_me=*/ false, coin_selection_params.m_effective_feerate);
+        std::map<COutPoint, CAmount> map_of_bump_fees{wallet.chain().CalculateIndividualBumpFees(list_selected, coin_selection_params.m_effective_feerate)};
         output.ApplyBumpFee(map_of_bump_fees.at(output.outpoint));
         result.Insert(output, coin_selection_params.m_subtract_fee_outputs);
     }
```

murchandamus replied:
Maybe I’m misunderstanding what you propose, but `CalculateIndividualBumpFees(…)` gets the bump fees for all elements of `coin_control.ListSelected()` at once, and this would move it into the for-loop, so we’d get the individual bump fees for all `n` elements `n` times for n² effort.

brunoerg commented on src/wallet/spend.cpp:
```diff
@@ -348,6 +353,16 @@ CoinsResult AvailableCoins(const CWallet& wallet,
     }
 
+    if (feerate.has_value()) {
+        std::map<COutPoint, CAmount> map_of_bump_fees = wallet.chain().CalculateIndividualBumpFees(outpoints, feerate.value());
```
In 35a35fa989f14406f8218c20c4e4e732cee0bd6c: Shouldn't we ensure that `map_of_bump_fees` is not empty?

murchandamus replied:
I guess technically the `CalculateIndividualBumpFee(…)` call could return an empty map if the feerate changed during the call. I put an assert, and ran all the tests fine. In the worst case, it would fail right below where it applies the bump fees, so I’m not sure an assert would change much.

brunoerg replied:
Makes sense, just wondering about it.

murchandamus replied:
I see, thanks for asking!

S3RK commented on src/wallet/test/coinselector_tests.cpp:
```diff
@@ -827,100 +829,129 @@ BOOST_AUTO_TEST_CASE(SelectCoins_test)
-    add_coin(1 * COIN, 1, selection, fee, fee + large_fee_diff);
-    add_coin(2 * COIN, 2, selection, fee, fee + large_fee_diff);
-    BOOST_CHECK_EQUAL(target_waste2, GetSelectionWaste(selection, change_cost, target));
+    // The following tests that the waste is calculated correction in various scenarios
```
nit: correction -> correctly

murchandamus replied:
Done!

S3RK commented on src/wallet/coinselection.cpp:
```diff
@@ -459,7 +459,7 @@ CAmount SelectionResult::GetSelectionWaste(CAmount change_cost, CAmount target,
     for (const auto& coin_ptr : m_selected_inputs) {
         const COutput& coin = *coin_ptr;
-        waste += coin.GetFee() - coin.long_term_fee;
+        waste += coin.GetFee() + coin.ancestor_bump_fees - coin.long_term_fee;
```
nit: no need to add it here, only to remove in the next commit

murchandamus replied:
If I don’t add it here, I don’t think the tests would pass.

murchandamus replied:
I pulled the part of the waste test that did not deal with the group discount into this commit and actually realized this line was incorrect (and then correctly removed in the next commit). Thank you for pointing this out.

S3RK commented on src/wallet/spend.cpp:
```diff
@@ -544,13 +544,13 @@ FilteredOutputGroups GroupOutputs(const CWallet& wallet,
 static bool HasErrorMsg(const util::Result<SelectionResult>& res) { return !util::ErrorString(res).empty(); }
 
-util::Result<SelectionResult> AttemptSelection(const CAmount& nTargetValue, OutputGroupTypeMap& groups,
+util::Result<SelectionResult> AttemptSelection(const CWallet& wallet, const CAmount& nTargetValue, OutputGroupTypeMap& groups,
```
nit: could you pass only the `ChainInterface` here? Let's not pass the whole wallet around

S3RK commented on src/wallet/coinselection.cpp:
```diff
@@ -470,6 +471,7 @@ CAmount SelectionResult::GetSelectionWaste(CAmount change_cost, CAmount target,
         waste += change_cost;
     } else {
         // When we are not making change (change_cost == 0), consider the excess we are throwing away to fees
+        CAmount selected_effective_value = use_effective_value ? GetSelectedEffectiveValue() : GetSelectedValue();
```
It seems to me in the case with the excess `bump_fee_group_discount` will be canceled.

first you subtract the discount `waste -= bump_fee_group_discount;` but then `GetSelectedEffectiveValue()` adds it back `bump_fee_group_discount`

I think you only need to do `waste -= bump_fee_group_discount;` when there is a change, but I'm not absolutely sure.

Could you add a test for the waste calculation with bump_fee and excess?

murchandamus replied:
I’ve added a test. I think the canceling out of the `bump_fee_group_discount` is expected: both the bump fees and the excess count towards waste in full, so whether we internally attribute the additional fees to the group discount or the excess should not matter for the overall waste score.

S3RK reviewed (commented):
Started re-review. Besides few nits, I have a doubt about waste calculation with bump fee and excess. Could you take a look, please?

murchandamus reviewed (commented):
Addressed @s3rk’s review

murchandamus committed:
Amend bumpfee for inputs with overlapping ancestry

At the end of coin selection reduce the fees by the difference between
the individual bump fee estimates and the collective bump fee estimate.

murchandamus reviewed (commented):
I pulled the part of the waste test that dealt with bump fees without group discount into the earlier commit and fixed an issue in the waste calculation in that commit. The final code is unchanged.

murchandamus commented:
I had forgotten to update the `AttemptSelection` call in `src/bench/coin_selection.cpp` to the new function header that takes the chain instead of the wallet.

I should have now addressed all feedback from @brunoerg and @S3RK, and be ready for review again.

murchandamus commented:
@S3RK noticed that `SelectionResult::GetBumpFeeDiscount()` was now obsolete and no longer used. (AFAIR, this would have been a bi-product of moving GetSelectionWaste into SelectionResult.)

I removed the obsolete method.

S3RK commented:
ACK f18f9ef4d31c70e2d71ab90a24511692821418c3

Great work! This PR not only makes our wallet work properly with unconfirmed inputs, but it also improved the factoring of  the code. Happy to see waste calculations as a member function of `SelectionResult`

murchandamus commented:
@t-bast, @ismaelsadeeq, @achow101: [Changes since 1b2fb54d5c18579e013b1dfd061c2b29d8cffdc2](https://github.com/bitcoin/bitcoin/compare/1b2fb54..f18f9ef4d31c70e2d71ab90a24511692821418c3)

brunoerg reviewed (approved):
crACK f18f9ef4d31c70e2d71ab90a24511692821418c3

code lgtm, nice improvement!

ismaelsadeeq commented:
ACK f18f9ef4d31c70e2d71ab90a24511692821418c3

t-bast reviewed (approved):
ACK https://github.com/bitcoin/bitcoin/pull/26152/commits/f18f9ef4d31c70e2d71ab90a24511692821418c3, I reviewed the latest changes and run e2e tests against eclair, everything looks good :+1:

achow101 commented:
ACK f18f9ef4d31c70e2d71ab90a24511692821418c3
